File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxPathSum = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        maxPathSumMathod(root);
        return maxPathSum;
    }
    public int maxPathSumMathod(TreeNode root){
        if(root==null) return 0;
        int left = maxPathSumMathod(root.left);
        int right = maxPathSumMathod(root.right);
        left = Math.max(left, 0);
        right = Math.max(right, 0);
        maxPathSum = Math.max(maxPathSum, left+right+root.val);
        return root.val + Math.max(left, right);
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int maxVal=INT_MIN;
public:
    int maxPathSum(TreeNode* root) {
        if(root==NULL) return 0;
        dfs(root);
        return maxVal;
    }
    int dfs(TreeNode* root){
        if(root==NULL) return 0;
        int leftSum = dfs(root->left);
        int rightSum = dfs(root->right);
        leftSum = max(0, leftSum);
        rightSum = max(0, rightSum);
        maxVal = max(maxVal, root->val+leftSum+rightSum);
        return root->val+max(leftSum,rightSum);
    }
};